router/core/src/main/scala/com/twitter/finagle/buoyant/DstBindingFactory.scala

```diff
@@ -126,12 +133,16 @@ object DstBindingFactory {
     boundMk: Mk[Dst.Bound, Req, Rsp] = Mk.identity[Dst.Bound, Req, Rsp],
     namer: NameInterpreter = DefaultInterpreter,
     statsReceiver: StatsReceiver = DefaultStatsReceiver,
-    capacity: Capacity = Capacity.default
-  ) extends DstBindingFactory[Req, Rsp] {
+    capacity: Capacity = Capacity.default,
+    bindingTimeout: BindingTimeout = BindingTimeout.default
+  )(implicit timer: Timer = DefaultTimer.twitter) extends DstBindingFactory[Req, Rsp] {
     private[this]type Cache[Key] = XXX_ServiceFactoryCache[Key, Req, Rsp]
 
     def apply(dst: Dst, conn: ClientConnection): Future[Service[Req, Rsp]] = dst match {
-      case path: Dst.Path => pathCache(path, conn)
+      case path: Dst.Path => pathCache(path, conn).raiseWithin(
+        bindingTimeout.timeout,
+        new TimeoutException(s"Binding ${path.path.show} exceeded ${bindingTimeout.timeout}") with NoStacktrace
+      )
```

Comment:
Yeah, my concern is that we're creating a new throwable on every lookup regardless of whether a timeout is set or the exception is actually thrown.  Perhaps use [RequestTimeoutException](https://github.com/twitter/finagle/blob/develop/finagle-core/src/main/scala/com/twitter/finagle/Exceptions.scala#L108) which is mixes in TimeoutException with NoStacktrace?